### PR TITLE
[SYCL-MLIR][SYCLToLLVM] Add option `indexBitwidth`

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Address code review comment https://github.com/intel/llvm/pull/8703#discussion_r1144868678.

### CODE REVIEW

File: mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td [hunk 1]
```diff
@@ -380,7 +380,7 @@ def SYCLGetIDResult : AnyTypeOf<[I64, SYCL_IDType]>;
 def SYCLGetRangeResult : AnyTypeOf<[I64, SYCL_RangeType]>;
 def SYCLIndexSpaceGetIDResult : AnyTypeOf<[Index, SYCL_IDType]>;
 def SYCLIndexSpaceGetRangeResult : AnyTypeOf<[Index, SYCL_RangeType]>;
-def SYCLGetResult : AnyTypeOf<[I64, MemRefOf<[I64]>]>;
+def SYCLGetResult : AnyTypeOf<[I64, MemRefOf<[I64]>, I32, MemRefOf<[I32]>]>;
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/IR/SYCLOps.td:383]: This should be `AnyTypeOf<[Index, MemRefOf<[Index]>]>` IMO, but that would require work in codegen. Can you put a comment here?

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 2]
```diff
@@ -2225,10 +2229,11 @@ LogicalResult ConvertSYCLToLLVMPass::convertToSPIRV() {
 
   auto &context = getContext();
   auto module = getOperation();
+  auto &bitIndex = this->indexBitwidth;
```
Comment (victor-eds) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2232]:
```suggestion
  const auto bitIndex = this->indexBitwidth;
```

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 3]
```diff
@@ -2225,10 +2229,11 @@ LogicalResult ConvertSYCLToLLVMPass::convertToSPIRV() {
...
 
   const auto res = failure(
       module
-          .walk([&context](gpu::GPUModuleOp gpuModule) {
+          .walk([&context, &bitIndex](gpu::GPUModuleOp gpuModule) {
```
Comment (victor-eds) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2236]:
```suggestion
          .walk([&context, bitIndex](gpu::GPUModuleOp gpuModule) {
```

**Review by victor-eds — COMMENTED** [review]
LGTM after NITs :+1:

File: mlir-sycl/include/mlir/Conversion/SYCLPasses.td [hunk 4]
```diff
@@ -29,6 +29,11 @@ def ConvertSYCLToLLVM : Pass<"convert-sycl-to-llvm", "ModuleOp"> {
       "spirv::SPIRVDialect",
       "vector::VectorDialect"
   ];
+  let options = [
+    Option<"indexBitwidth", "index-bitwidth", "unsigned",
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply mlir-sycl/include/mlir/Conversion/SYCLPasses.td:33]: I would like to use the same option name as other passes e.g., ArithToLLVM, ControlFlowToLLVM, FuncToLLVM, etc.

**Review by victor-eds — APPROVED** [review]

File: mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp [hunk 5]
```diff
@@ -2237,8 +2242,9 @@ LogicalResult ConvertSYCLToLLVMPass::convertToSPIRV() {
             auto targetAttr = spirv::lookupTargetEnvOrDefault(gpuModule);
             auto target = SPIRVConversionTarget::get(targetAttr);
             SPIRVConversionOptions options;
-            // TODO: Add 32 bits support.
-            options.use64bitIndex = true;
+            options.use64bitIndex =
+                (bitIndex == kDeriveIndexBitwidthFromDataLayout ||
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2246]: > So what happen if the data layout indicates that the pointer width is 32 ? We will use 64 here which is not correct.

Then `bitIndex` is 32, and `options.use64bitIndex` is `false`.
Notice that passes are added like
```
  ConvertFuncToLLVMPassOptions funcToLLVMOptions{};
  funcToLLVMOptions.indexBitwidth =
      DataLayout(module).getTypeSizeInBits(IndexType::get(module.getContext()));
  passManager.addPass(createConvertFuncToLLVMPass(funcToLLVMOptions));
```

[the remark this answers is not in the document]
Comment (etiotto) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp:2246]: OK, the code snippet in the last comment shows that indexBitwidth gets the size of the the pointer index type from the module data layout which is what I wanted to ensure was happening in my question. So looks good.

**Review by etiotto — APPROVED** [review]

**Pull request merged**